Add blog post on probability (#93)
leanprover-community/blog: posts/basic-probabliity-in-lean.md

```lean
variable {Ω : Type*} [MeasurableSpace Ω] [TopologicalSpace Ω] [BorelSpace Ω]
```

For properties related to conditional distributions, it is often convenient or necessary to work in a standard Borel space (a measurable space arising as the Borel sets of some Polish topology). See the `StandardBorelSpace` typeclass.
Note that a countable discrete measurable space is a standard Borel space, so there is no need to worry about that typeclass when doing discrete probability.

## Various probability definitions

This section contains pointers to the Mathlib definitions for several probability objects.
That list might be out of date when you read this! Look around in the [documentation](https://leanprover-community.github.io/mathlib4_docs/).

### CDF, pdf, variance, moments

Here is a list of common concepts about real probability distributions.

- Probability density function of a random variable `X : Ω → E` in a space with measure `P : Measure Ω`, with respect to measure `Q : Measure E`: `pdf X P Q`.
- Cumulative distribution function of `P : Measure ℝ`: `cdf P`. This is a `StieltjesFunction`, a monotone right continuous real function.

> - Expectation of `X` under `P`: `P[X]`.

Reviewer: Could you expand on this a bit? I was trying some probability theory earlier this week, for the first time in at least a year, and the expectation notation was the only part I couldn't figure out. More specifically, have `P : Measure R` with `[MeasurableSpace R]`, and `[IsProbabilityMeasure P]`. I also have a function `R -> ENNReal`, which is measurable, and I'd like its average. I ended up using `lintegral` for this, but I expected (no pun intended) one of the expectation notations to work... 

So, could you clarify the types of `X` and `P` here, and which assumptions are needed for the notation to work?

Author: Thanks for the feedback. All the current expectation notations correspond to Bochner integrals and not Lebesgue, so they won't work for ENNReal.

I talk about the expectation in a bit more detail earlier in the post (in the random variable section): I'll add a remark over there about the two types of integrals and the assumptions on the spaces.

Author: I added a remark.

Reviewer: Ah perfect, thanks!

```lean
example (P : Measure ℝ) (s : Set ℝ) : ℝ≥0∞ := P s
```
The probability of the event `s` is thus `P s`.
The type `ℝ≥0∞` represents the nonnegative reals and infinity: the measure of a set is a nonnegative real number which in general may be infinite.
If `P` is a probability measure, it actually takes only values up to 1.
The tactic `simp` knows that a probability measure is finite and will use the lemmas `measure_ne_top` or `measure_lt_top` to prove that `P s ≠ ∞` or `P s < ∞`.

The operations on `ℝ≥0∞` are not as nicely behaved as on `ℝ`: `ℝ≥0∞` is not a ring. For example, subtraction truncates to zero.
If one finds that lemma `lemma_name` used to transform an equation does not apply to `ℝ≥0∞`, a good thing to try is to find a lemma named like `ENNReal.lemma_name_of_something` and use that instead (it will typically require that one variable is not infinite).

For many lemmas to apply, the set `s` will need to be a measurable set. The way to express that is `MeasurableSet s`.

## Random variables

A random variable is a measurable function from a measurable space to another.
```lean
variable {Ω : Type*} [MeasurableSpace Ω] {X : Ω → ℝ} (hX : Measurable X)
```

> In that code we defined a random variable `X` from the measurable space `Ω` to `ℝ` (for which the typeclass inference system finds a measurable space instance). `hX` states that `X` is measurable, which is necessary for most manipulations.

Reviewer: ```suggestion
In that code we defined a random variable `X` from the measurable space `Ω` to `ℝ` (for which the typeclass inference system finds a measurable space instance). The assumption `hX` states that `X` is measurable, which is necessary for most manipulations.
```

> If we define a measure `P` on `Ω`, we can talk about the law or distribution of `X`. It is by definition the map of the measure `P` by `X`, `P.map X`. There is no specific notation for that law.

Reviewer: The definitions `map` and `comap` are mentioned a couple of times throughout this post, I wonder if it would be useful to say something more about them? Or at least, the way in which they're useful with respect to probability theory

Author: I added two sentences about `Measure.map` and `MeasurableSpace.comap` (which are the two definitions used here) where they are first mentioned. I don't think there is much more to say about those while staying at the introductory level of the post.
Tell me if you had something else in mind.

To say that `X` is Gaussian with mean 0 and variance 1, write `P.map X = gaussianReal 0 1`.

The expectation of `X` is the integral of that function against the measure `P`, written `∫ ω, X ω ∂P`.
The notation `P[X]` is shorthand for that expectation. In a `MeasureSpace`, we can further use the notation `𝔼[X]`.

Remark: there are two types of integrals in Mathlib, Bochner integrals and Lebesgue integrals.
The expectation notations stand for the Bochner integral, which is defined for `X : Ω → E` with `E` a normed space over `ℝ` (`[NormedAddCommGroup E] [NormedSpace ℝ E]`).
They don't work for `Y : Ω → ℝ≥0∞` since `ℝ≥0∞` is not a normed space, but those functions can be integrated with the Lebesgue integral: `∫⁻ ω, Y ω ∂P`.
There is no expectation notation for the Lebesgue integral.

## Discrete probability

In discrete probability, measurability is not an issue: every set and every function are measurable.
The typeclass `[DiscreteMeasurableSpace Ω]` signals that every set of `Ω` is measurable and the lemma `MeasurableSet.of_discrete` provides a proof of measurability.
To obtain measurability of a function from `Ω`, use `Measurable.of_discrete`.

> Any countable type with measurable singletons is a `DiscreteMeasurableSpace`. For example `ℕ` or `Fin n`.

Reviewer: ```suggestion
Any countable type with measurable singletons is a `DiscreteMeasurableSpace`, for example `ℕ` or `Fin n`.
```

> A way to define a probability measure on a discrete space `Ω` is to use the type `PMF Ω`, which stands for probability mass function.

Reviewer: This describes a convenient way to define a probability measure on a discrete space, but it doesn't mention if `PMF  Ω` is something that should be used as an assumption in general. For instance, let's say I want to state a theorem of the form "given a probability measure P on a discrete finite space...". Should I state the theorem over `P : PMF Ω`, or instead over `[DiscreteMeasurableSpace Ω]` with `P : Measure Ω` and `[IsProbabilityMeasure Ω]`? I'm guessing the latter, but I genuinely don't know the answer!

Author: To be honest I did not use PMF much, so I am unsure about it. I wrote that it was "a way to define" and not "a convenient way to define" because I am not sure about that, but at the same time I felt like I had to include a word about PMF.

Several other repositories opted for not using PMF (because bundling the summability assumption can be inconvenient, as far as I can tell). For example https://github.com/google-deepmind/debate and https://github.com/leanprover/SampCert/tree/main.

My feeling is that stating the theorem for measures will integrate better with other things in Mathlib, but I lack the practice with PMF to be assertive about that.

Reviewer: Indeed, this was also the choice I made in my situation! I understand the hesitation to be assertive here, but I wonder if this might be exactly the right place in which to create a convention for how these things are meant to be used

Author: Ok let's use that opportunity to make a choice. I added a sentence about preferring Measure over PMF in lemmas.

- Cumulative distribution function of `P : Measure ℝ`: `cdf P`. This is a `StieltjesFunction`, a monotone right continuous real function.
- Expectation of `X` under `P`: `P[X]`.
- Variance: `variance X P`.
- Moment of order `p`: `moment X p P`.
- Central moment of order `p`: `centralMoment X p P`.
- Moment generating function of `X` under `P` at `t : ℝ`: `mgf X P t`.
- Cumulant generating function: `cgf X P t`.

### Known probability distributions

The Probability/Distributions folder of Mathlib contains several common probability laws: Exponential, Gamma, Gaussian (only in `ℝ`), Geometric, Pareto, Poisson, Uniform.
Other measures that can be used to build probability distributions are defined elsewhere, in the MeasureTheory folder: counting measure, Dirac, Lebesgue measure.

For example, the Gaussian distribution on the real line with mean `μ` and variance `v` is a `Measure ℝ`:
```lean
def gaussianReal (μ : ℝ) (v : ℝ≥0) : Measure ℝ := -- omitted
```
The definition is followed by an instance `IsProbabilityMeasure (gaussianReal μ v)` that states that this is a probability measure. It is defined as the Dirac distribution for `v = 0`.

> The file where the Gaussian is defined also contains properties of its p.d.f.

Reviewer: ```suggestion
The file where the Gaussian is defined also contains properties of its probability density function.
```

As another example, to take the uniform distribution on a finite set `s : Set Ω`, use `uniformOn s : Measure Ω`.

### Identically distributed

`IdentDistrib X Y P Q` (or `IdentDistrib X Y` in `MeasureSpace`) states that `X` and `Y` are identically distributed.

### Conditioning

#### Conditional probability

The probability of a set `s` conditioned on another set `t` for the measure `P` is `P[s|t]`, which equals `P (s ∩ t) / P t`.
Conditioning on `t` defines a new measure named `cond P t`, denoted by `P[|t]`. With that notation, `P[s|t] = P[|t] s`.

For `X : Ω → E` and `x : E`, we write `P[|X ← x]` for the probability measure conditioned on `X = x`. It is notation for `P[|X ⁻¹' {x}]`.
This is meaningful only if `X = x` has non-zero probability, so that definition is mostly useful for discrete probability.

#### Conditional expectation

> The conditional expectation of a random variable `Y : Ω → F` given a sigma-algebra `m` of `Ω` is `condexp m P Y`, with notation `P[Y | m]`. This is a random variable `Ω → F`.

Reviewer 2: ```suggestion
The conditional expectation of a random variable `Y : Ω → F` given a sigma-algebra `m` of `Ω` is `condexp m P Y`, with notation `P[Y | m]`. This is a random variable `Ω → F`, and it is `m`-measurable.
```

The conditional expectation of `Y` given `X : Ω → E` can be obtained with `P[Y | mE.comap X]`, in which `mE : MeasurableSpace E` is the sigma-algebra on `E`.

We have special notation for the conditional expectation of the real indicator of a set `s : Set Ω`: `P⟦s | m⟧ : Ω → ℝ`.

#### Conditional probability distribution

The regular conditional probability distribution of `Y : Ω → F` given `X : Ω → E`, for `F` standard Borel, is denoted by `condDistrib Y X P` (for a measure `P` on `Ω`).

> This is a Markov kernel from `E` to `F` (see further down) such that for all measurable sets `s` of `F`, for `P`-almost every `ω : Ω `, `condDistrib Y X P (X ω) s` is equal to `P⟦Y ⁻¹' s|mE.comap X⟧` (up to a mismatch in types: `ℝ≥0∞` for `condDistrib` versus `ℝ` for the conditional expectation of the indicator).

Reviewer 2: ```suggestion
This is a Markov kernel from `E` to `F` (see further down) such that for all measurable sets `s` of `F`, for `P`-almost every `ω : Ω`, `condDistrib Y X P (X ω) s` is equal to `P⟦Y ⁻¹' s|mE.comap X⟧` (up to a mismatch in types: `ℝ≥0∞` for `condDistrib` versus `ℝ` for the conditional expectation of the indicator).
```

> 

Reviewer 2: Add a comment somewhere that this only makes sense in standard Borel spaces? (At other places of the blog post also, maybe?)

Author: There is "for `F` standard Borel" in the first sentence of that paragraph, so I alluded to it a bit already. Do you want me to be more explicit about it?

Author: I added a remark about standard Borel in the conditional independence section.

Reviewer 2: You read my mind: this remark came to me while I was reading the paragraph on conditional independence (where standard Borel was not mentioned), but then I thought I should mention it at the first place it was relevant, and I didn't notice it was indeed mentioned there. Sorry!

### Independence

Mathlib has several definitions for independence. We can talk about independence of random variables, of sets, of sigma-algebras.
The definitions also differ depending on whether we consider only two random variables of an indexed family.
Finally, there are also conditional variants of all those definitions.

#### Unconditional independence

Two independent random variables can be defined like this:
```lean
variable {Ω : Type*} [MeasurableSpace Ω] {P : Measure Ω}
  {X : Ω → ℝ} {Y : Ω → ℕ} (hX : Measurable X) (hY : Measurable Y)
  (hXY : IndepFun X Y P)
```
On a measure space, we can write `IndepFun X Y` without the measure argument.

> For family `X : ι → Ω → ℝ` of independent random variables, use `iIndepFun`.

Reviewer: ```suggestion
For a family `X : ι → Ω → ℝ` of independent random variables, use `iIndepFun`.
```

> For family `X : ι → Ω → ℝ` of independent random variables, use `iIndepFun`.

Reviewer 2: ```suggestion
For a family `X : ι → Ω → ℝ` of independent random variables, use `iIndepFun`.
```

> For sets, use `IndepSet` (two sets) and `iIndepSet` (family of sets). For sigma-algebras, `Indep` and `iIndep`.

Reviewer: ```suggestion
To express independence of sets, use `IndepSet` (two sets) and `iIndepSet` (family of sets). For sigma-algebras, `Indep` and `iIndep`.
```

#### Conditional independence

> The way to write that two random variables `X : Ω → E` and `Y : Ω → F` are independent conditionally on a sub-sigma-algebra `m` with respect to the measure `P` is `CondIndepFun m hm X Y P`, in which `hm : m ≤ mΩ` is a proof that `m` is a sub-sigma-algebra of the sigma-algebra `mΩ` of `Ω`. We would omit the measure argument in a `MeasureSpace`.

Reviewer: ```suggestion
The way to write that two random variables `X : Ω → E` and `Y : Ω → F` are conditionally independent on a sub-sigma-algebra `m` with respect to the measure `P` is `CondIndepFun m hm X Y P`, in which `hm : m ≤ mΩ` is a proof that `m` is a sub-sigma-algebra of the sigma-algebra `mΩ` of `Ω`. We would omit the measure argument in a `MeasureSpace`.
```
This reads slightly better to me

Author: I replaced it with "conditionally independent given a sub-sigma-algebra `m`", which is the way it's spelled on wikipedia.

To write that `X : Ω → E` and `Y : Ω → F` are independent conditionally on a third random variable `Z : Ω → G` (with measurability assumption `hZ : Measurable Z`), write that they are independent given the sigma-algebra generated by `Z`.
That is, for `mG : MeasurableSpace G` the sigma-algebra on `G`, write `CondIndepFun (mG.comap Z) hZ.comap_le X Y P`.
As of writing this blog post, there is no shorter spelling of that fact.

> For families of function, use `iCondIndepFun`.

Reviewer: ```suggestion
For families of functions, use `iCondIndepFun`.
```

## Probability of events

An event is a measurable set: there is no special event definition in Mathlib.
The probability of that event is the measure of the set.
A `Measure` can be applied to a set like a function and returns a value in `ENNReal` (denoted by `ℝ≥0∞`, available after `open scoped ENNReal`).
```lean
example (P : Measure ℝ) (s : Set ℝ) : ℝ≥0∞ := P s
```
The probability of the event `s` is thus `P s`.
The type `ℝ≥0∞` represents the nonnegative reals and infinity: the measure of a set is a nonnegative real number which in general may be infinite.
If `P` is a probability measure, it actually takes only values up to 1.
The tactic `simp` knows that a probability measure is finite and will use the lemmas `measure_ne_top` or `measure_lt_top` to prove that `P s ≠ ∞` or `P s < ∞`.

The operations on `ℝ≥0∞` are not as nicely behaved as on `ℝ`: `ℝ≥0∞` is not a ring. For example, subtraction truncates to zero.
If one finds that lemma `lemma_name` used to transform an equation does not apply to `ℝ≥0∞`, a good thing to try is to find a lemma named like `ENNReal.lemma_name_of_something` and use that instead (it will typically require that one variable is not infinite).

> 

Reviewer 2: I'm a little bit sad that we can't mention `P.real` there, that was written for PFR but never upstreamed because it was blocked by the upstreaming of the `Finiteness` tactic. @YaelDillies , didn't you mention at some point upstreaming it was in your plans?

Reviewer 2: Definitely not! To me the blog post looks ready, in fact. Maybe it's a good idea to update it in the future when new notation or new possibilities appear, but better post it as soon as possible.

Reviewer 3: You might have noticed from the queue that I have started seriously upstreaming PFR. I haven't gotten to `finiteness` yet, but it's only a matter of days.

For many lemmas to apply, the set `s` will need to be a measurable set. The way to express that is `MeasurableSet s`.

## Random variables

A random variable is a measurable function from a measurable space to another.
```lean
variable {Ω : Type*} [MeasurableSpace Ω] {X : Ω → ℝ} (hX : Measurable X)
```
In that code we defined a random variable `X` from the measurable space `Ω` to `ℝ` (for which the typeclass inference system finds a measurable space instance). The assumption `hX` states that `X` is measurable, which is necessary for most manipulations.

If we define a measure `P` on `Ω`, we can talk about the law or distribution of a random variable `X : Ω → E`.

> The law of `X` is a measure on `E`, with value `P (X ⁻¹' s)` on any measurable set `s` of `E`.
> This is the map of the measure `P` by `X`, `Measure.map X P` or more succinctly `P.map X`.

Reviewer: ```suggestion
The law of `X` is a measure on `E`, with value `P (X ⁻¹' s)` on any measurable set `s` of `E`.
This is how we define the map of the measure `P` by `X`, `Measure.map X P` or more succinctly `P.map X`.
```
optional, but I think it's helpful to say something along these lines

We can write that it is a martingale with respect to `ℱ` and the measure `P : Measure Ω` with `Martingale X ℱ P`.

A stopping time with respect to a filtration `ℱ` is a function `τ : Ω → ι` such that for all `i`, the preimage of `{j | j ≤ i}` along `τ` is measurable with respect to `ℱ i`. 
`IsStoppingTime ℱ τ` states that `τ` is a stopping time.

Remark: there is an issue with the current definition of stopping times, which is that if the index set of the filtration `ι` is the set of natural numbers `ℕ` then `τ` has to take values in `ℕ`. In particular it cannot be infinite.
That can cause friction with informal stopping times which are often understood to be infinite when they are not well defined.

### Transition kernels

A transition kernel `κ` from a measurable space `E` to a measurable space `F` is a measurable function from `E` to `Measure F`. That is, for every measurable set `s` of `F`, the function `fun x ↦ κ x s` is measurable.
Transition kernels are represented in Mathlib by the type `Kernel E F`.
A kernel such that all measures in its image are probability measures is called a Markov kernel. This is denoted by `IsMarkovKernel κ`.

> There are other typeclasses for kernels that are finite (`IsFiniteKernel`) or s-finite (countable sum of finite, `IsSFiniteKernel`).

Reviewer 3: None of these names are fully qualified. What about either fully qualifying the names or turning the code snippets into links to the docs?
